### PR TITLE
discover: init at 5.15.3

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Motivation for this change

Is this worth merging considering discover doesn't actually do anything on NixOS?

Cc: @bkchr @adisbladis @ttuegel 

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by bkchr** [comment]
I think it is not useful, because everything you install will not be runable. So, I would not add the discover package.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -15009,7 +15009,7 @@ with pkgs;
   inherit (gnome3) evince;
   evolution_data_server = gnome3.evolution_data_server;
 
-  keepass = callPackage ../applications/misc/keepass { 
+  keepass = callPackage ../applications/misc/keepass {
```
Comment (dtzWill) [inline pkgs/top-level/all-packages.nix]: I'm all for removing trailing whitespace, but probably best to do so separately.

File: pkgs/desktops/plasma-5/discover.nix [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+  karchive, kconfig, kcrash, kdbusaddons, kdeclarative, kio, kitemmodels, kxmlgui
+}:
+
+mkDerivation {
+  name = "discover";
+  nativeBuildInputs = [ extra-cmake-modules gettext kdoctools python ];
+  buildInputs = [
+    qtquickcontrols2
+    karchive kconfig kcrash kdbusaddons kdeclarative kio kitemmodels kxmlgui
+  ];
```
Comment (dtzWill) [inline pkgs/desktops/plasma-5/discover.nix]: meta, please!

**Review by dtzWill — CHANGES_REQUESTED** [review]
Minor comments, if we end up wanting this.

**Comment by dtzWill** [comment]
And for those wondering what "discover" is supposed to do anyway: https://userbase.kde.org/Discover

After reading that I would agree it doesn't seem to be a good idea to inlcude.  Does existing software cleanly deal with its omission (no do-nothing buttons or borked context menu entries...)?

**Comment by ttuegel** [comment]
I agree, I don't think this adds any value for our users.

> Does existing software cleanly deal with its omission (no do-nothing buttons or borked context menu entries...)?

AFAIK, everything works correctly without it. And if not, that's upstream's problem for not checking dependencies 😉

**Comment by peterhoeg** [comment]
Discover supposedly supports flatpak (and snap?) as a mechanism to install software which could be compatible with NixOS albeit far from the preferred option for getting software installed.

**Comment by bkchr** [comment]
I think you are right, but are they only ship flatpaks/snaps with Discover?

**Comment by peterhoeg** [comment]
> but are they only ship flatpaks/snaps with Discover?

Discover can also be used to install/upgrade distribution packages which doesn't apply in to our situation.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (peterhoeg) [reply pkgs/top-level/all-packages.nix]: emacs trims it automatically but I could split it out of course.

File: pkgs/desktops/plasma-5/discover.nix (hunk 2 above)
Comment (peterhoeg) [reply pkgs/desktops/plasma-5/discover.nix]: None of the software distributed as part of KDE Applications carry meta info.

**Comment by peterhoeg** [comment]
So I haven't given up completely on this yet. 

Discover can be used to handle things obtained through "Get New Hot Stuff" such as icons, wallpapers, plasma widgets and so on. When (if?) we get flatpak support at some point, discover can be used for that too. And should anyone decide to write a nix-env backend that would be an option as well.

In my opinion, the "Get Hot New Stuff" use-case is sufficiently strong to include discover.

Note: The appstream bits are included in #35308.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (dtzWill) [reply pkgs/top-level/all-packages.nix]: Your call, folks don't seem to be particular about this.

File: pkgs/desktops/plasma-5/discover.nix (hunk 2 above)
Comment (dtzWill) [reply pkgs/desktops/plasma-5/discover.nix]: Oh? Interesting, well nevermind then :1+.

**Comment by jtojnar** [comment]
We will have [packagekit-nix](https://github.com/NixOS/nixpkgs/issues/21230) eventually. And [flatpak support](https://github.com/NixOS/nixpkgs/pull/33371) is only missing tests in order to be merged.

**Comment by peterhoeg** [comment]
The reasons for not merging discover are disappearing fast! Thanks for your work on flatpak - looking forward to that.

**Comment by peterhoeg** [comment]
We're getting closer although the flatpak bits aren't quite working yet.

**Comment by jtojnar** [comment]
Any progress?

**Comment by peterhoeg** [comment]
Not quite. It's currently blowing up complaining about packagekit but I haven't had the time to look at it.

**Comment by peterhoeg** [comment]
This PR alone allows installing application and plasma addons (not applications) but requires #60146 in order to not complain like crazy.

While flatpak support is compiled in, it doesn't work yet but I don't think that's a show stopper for merging as is.

**Review by ttuegel — APPROVED** [review]

**Comment by dtzWill** [comment]
Using somewhat different tree[1] I can run this and successfully installed a flatpak app "kstars" and it runs successfully!

[1]
No clue if any of this matters, but JIC:
* 5.15.4 (https://github.com/NixOS/nixpkgs/pull/57520#issuecomment-480420357 I guess) 
* newer (dev) flatpak (needs new user and such), and enabled nixos flatpak service
* I have misc flatpak remotes from gnome and kde (although TBH they generally haven't worked well for me, but dunno)

Anyway! This seems good--it's a bit unpolished UI experience that seems generally not always handling gracefully some of the details of our environment.  Anyway I think that's arguably upstream issue and our portion of it can be iterated on IMO.

Didn't have the courage to see what happens if I install things like cursors or themes O:).

**Comment by peterhoeg** [comment]
Can you share more details on getting flatpak running? I have it enabled as well as the kde portal, but discover doesn't show any flatpaks.

**Pull request merged**